packages/app/src/Charcoal/App/ServiceProvider/EventServiceProvider.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+namespace Charcoal\App\ServiceProvider;
+
+use Charcoal\App\Event\EventDispatcher;
+use Charcoal\App\Event\EventListenerInterface;
+use Charcoal\Factory\FactoryInterface;
+use Charcoal\Factory\GenericFactory;
+use Pimple\Container;
+use Pimple\ServiceProviderInterface;
+use Psr\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * Event Service Provider. Configures and provides a PDO service to a container.
+ */
+class EventServiceProvider implements ServiceProviderInterface
+{
+    /**
+     * @param Container $container A service container.
+     * @return void
+     */
+    public function register(Container $container)
+    {
+        /**
+         * @param Container $container Pimple DI container.
+         * @return EventDispatcherInterface
+         */
+        $container['event/dispatcher'] = function (Container $container): EventDispatcherInterface {
+            $dispatcher = new EventDispatcher();
+            $dispatcher->setLogger($container['logger']);
+
+            foreach ($container['event/listeners'] as $event => $listeners) {
+                if (is_string($listeners)) {
+                    $listeners = [$listeners];
+                }
+
+                foreach ($listeners as $listener => $options) {
+                    $listener = $container['event/listener/factory']->create($listener);
+
+                    $priority = ($options['priority'] ?? 0);
+                    $once     = ($options['once'] ?? false);
+
+                    if ($once) {
+                        $dispatcher->subscribeOnceTo($event, $listener, $priority);
+                        continue;
+                    }
+
+                    $dispatcher->subscribeTo($event, $listener, $priority);
+                }
+            }
```

Comment:
I suggest moving the registration of predefined events to a method of the service provider.

```php
$dispatcher = new EventDispatcher();
$dispatcher->setLogger($container['logger']);

$this->registerEventListeners($dispatcher, $container['event/listeners']);

return $dispatcher;
```

Reply:
Done